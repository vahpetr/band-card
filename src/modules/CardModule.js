// @ts-check

import React from "react";
import { connect } from "react-redux";
import { BondHighstockComponent } from "../components/BondHighstockComponent";
// @ts-ignore
import logo from "../logo.svg";
import "../App.css";
import { bondDataFething } from "../actions";
import { ValueTypeSelectComponent } from "../components/ValueTypeSelectComponent";

export class CardModule extends React.Component {
  componentDidMount() {
    const { dispatch, card } = this.props;
    dispatch(bondDataFething(card.bond.filter));
  }

  render() {
    const { data, filter } = this.props.card.bond;

    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Bond card</h1>
        </header>
        <BondHighstockComponent data={data} filter={filter} />
        <ValueTypeSelectComponent
          defaultValue={filter.valueType}
          onChange={this.onBondDataFilterChange}
        />
      </div>
    );
  }

  onBondDataFilterChange = (valueType) => {
    this.props.dispatch(bondDataFething({
      valueType
    }));
  }
}

const mapStateToProps = state => {
  return {
    card: state.cardReducer
  };
};
const mapDispatchToProps = dispatch => ({ dispatch });

export const ConnectedCardModule = connect(
  mapStateToProps,
  mapDispatchToProps
)(CardModule);
