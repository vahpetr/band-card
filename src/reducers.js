// @ts-check

import { combineReducers } from "redux";
import {
  BOND_DATA_FETCHING,
  BOND_DATA_SUCCEED,
  BOND_DATA_FAILED
} from "./actions";
import { ValueType } from "./types";

/**
 * App initial state
 */
export const initialState = Object.freeze({
  bond: {
    filter: {
      isin: "RU000A0JU4L3",
      valueType: ValueType.PRICE
    },
    data: [],
    error: null
  }
});

/**
 * Card reducer
 */
export function cardReducer(state = initialState, action) {
  switch (action.type) {
    case BOND_DATA_FETCHING: {
      const newState = { ...state };
      const newFilter = { ...newState.bond.filter, ...action.payload };
      newState.bond.filter = newFilter;
      return newState;
    }
    case BOND_DATA_SUCCEED: {
      const newState = { ...state };
      newState.bond.data = action.payload.data;
      newState.bond.error = null;
      return newState;
    }
    case BOND_DATA_FAILED: {
      const newState = { ...state };
      newState.bond.data = [];
      newState.bond.error = action.payload;
      return newState;
    }
    default: {
      return state;
    }
  }
}

/**
 * App reducer. All combinded reducers
 */
export const appReducers = combineReducers({
  cardReducer
});
