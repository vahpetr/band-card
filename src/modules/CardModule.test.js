// shallow+sinon http://airbnb.io/enzyme/
// redux-mock-store https://github.com/dmitry-zaets/redux-mock-store

import React from "react";
import { configure, shallow, mount } from "enzyme";
import sinon from 'sinon';
import { shallowToJson } from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import configureStore from "redux-mock-store";
import { ConnectedCardModule, CardModule } from "./CardModule";
import { initialState } from "../reducers";

configure({ adapter: new Adapter() });

describe("<ConnectedCardModule />", () => {

  it("without crashing", () => {
    const mockStore = configureStore();
    const store = mockStore(initialState);
    const wrapper = shallow(<ConnectedCardModule store={store} />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe("<CardModule />", () => {
    it("renders without crashing", () => {
      var callback = sinon.fake();
      const wrapper = mount(<CardModule card={initialState} dispatch={callback} />);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  
  });
});
