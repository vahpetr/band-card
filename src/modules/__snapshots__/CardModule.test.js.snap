// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectedCardModule /> <CardModule /> renders without crashing 1`] = `
<CardModule
  card={
    Object {
      "bond": Object {
        "data": Array [],
        "error": null,
        "filter": Object {
          "isin": "RU000A0JU4L3",
          "valueType": "Price",
        },
      },
    }
  }
  dispatch={[Function]}
>
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
      <h1
        className="App-title"
      >
        Bond card
      </h1>
    </header>
    <BondHighstockComponent
      data={Array []}
      filter={
        Object {
          "isin": "RU000A0JU4L3",
          "valueType": "Price",
        }
      }
    >
      <div
        className="highstock"
      />
    </BondHighstockComponent>
    <ValueTypeSelectComponent
      defaultValue="Price"
      onChange={[Function]}
    >
      <select
        defaultValue="Price"
        onChange={[Function]}
      >
        <option
          key="Yield"
          value="Yield"
        >
          Yield
        </option>
        <option
          key="Spread"
          value="Spread"
        >
          Spread
        </option>
        <option
          key="Price"
          value="Price"
        >
          Price
        </option>
      </select>
    </ValueTypeSelectComponent>
  </div>
</CardModule>
`;

exports[`<ConnectedCardModule /> without crashing 1`] = `
<CardModule
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
